### PR TITLE
Tests: implement snapshot testing

### LINKED ISSUE #45: Rewrite view tests using snapshots
In #37 we had to remove view tests that checked the context being rendered because they were basically duplicating application code and their usefulness is rather questionable.

Instead of this, we will rather use snapshot testing: take a snapshot of the rendered/returned context and store it in a file, serialized. This human-readable output will be checked by developers for validity and it will be committed to version control. Test runs will read from the snapshot file and compare it to the output provided by the view being tested.

Changes in view contexts/payloads mean the snapshot files need to be reviewed and updated too.

### PR DESCRIPTION
**Pull request by julen**
~_Note: This is pretty much a proof of concept and a work in progress_~

The PR implements snapshot testing for views.

The concept is simple: request a view via a URL and dump the rendered context to a JSON file. A human needs to evaluate first whether the rendered context is valid or not.

Afterwards, the verification process compares the expected context with the context generated by the test run. If they match, the test passes; fails otherwise.

Fixes #45.

### CODE REVIEW

**Comment by julen** [comment]
The failure here happens due to the ordering (or lack thereof) of the `checks` key in the context data. The ordering is different when the entire suite is run as opposed to only a few tests. This means we need to ensure the strings we generate always produce the same sorting (and JSON encoder's `sort_keys` is not enough as demonstrated by this PR).

**Comment by julen** [comment]
> The failure here happens due to the ordering (or lack thereof) of the checks key in the context data

Actually, while this _might_ happen, something completely different is getting in the way here. If the entire test suite is run, the `checks` list provided for browse views contains 74 items. If only these new view tests are run, the `checks` list contains 43 items.

By disabling the qualitycheck-related tests in _tests/misc/checks.py_ and _tests/commands/checks.py_ I am able to trim the list down to 44 items (still an intruder hanging around). This smells to me like shared mutable state — unfortunately I haven't been able to find the root cause yet 😞.

**Comment by iafan** [comment]
> This means we need to ensure the strings we generate always produce the same sorting (and JSON encoder's sort_keys is not enough as demonstrated by this PR).

@julen from our follow-up discussion it seems that you've found the root cause; I just want to comment on the specific sentiment I quoted above. Sorting JSON keys should be perfectly enough to generate the correct stable tree representation that you can compare across runs. If the data in the tree is identical, sorted JSON is guaranteed to produce the same result. As you have already noticed, the problem was that the data was different.

**Comment by julen** [comment]
> As you have already noticed, the problem was that the data was different.

That's right — I couldn't see the forest for the trees 😎.

**Comment by julen** [comment]
Cool, the PR has been refactored already and now it's in a usable state, i.e. we can start adding tests on top of this base.

Note Travis still fails because of the TTK bug affecting the test view we added but all works fine locally (it would be great if you could test this too @iafan). I also tried to switch Travis to my TTK branch which has the fix but it ignores it and uses the PyPI package no matter what so I gave up. We can opt to `xfail` these specific view tests on Travis for the time being and move on.

**Comment by iafan** [comment]
I checked out the PR and ran just the relevant test (`py.test -k test_browse`) and it shows 28 failures for me. Errors seems to be identical to this:

```
_________ test_browse[member2-/language0/project0/subdir0/store4.po] __________
tests\views\browse.py:35: in test_browse
    assert_ctx_matches_snapshot(response.context, snapshot_name)
tests\views\snapshots.py:80: in assert_ctx_matches_snapshot
    assert reference_file.read() == ctx_data
E   assert '{\n  "announ...: "nobody"\n}' == '{\n  "announc...: "nobody"\n}'
E     Skipping 430 identical leading characters in diff, use -v to show
E     - "code": "dialogsizes",
E     + "code": "dollar_sign_closure_placeholders",
E             "is_critical": true,
E     -       "title": "Dialog sizes",
E     +       "title": "$ closure placeholders",
E     -       "url": "/language0/project0/translate/subdir0/store4.po#filter=checks&checks=dialogsizes"
E     ?                                                                                      ^  - ^^
E     +       "url": "/language0/project0/translate/subdir0/store4.po#filter=checks&checks=dollar_sign_closure_placeholders"
E     Detailed information truncated (568 more lines), use "-vv" to show
```

It seems like the "dialogsizes" check comes from the standard set of checks in TTK, while "dollar_sign_closure_placeholders" is the Evernote's one. Is there anything I need to configure before running the tests? Note that we don't care about default checks and don't want them in the tests. They need to be superseded with our own ones. The PR contains data for the default checks, however.

**Comment by julen** [comment]
I bet your local settings have a custom value for `POOTLE_QUALITY_CHECKER`; if you comment it the tests should pass. Regardless of which checks we are using ourselves, I wanted to focus on making it work with the current defaults (as it is a _supported_ option for the time being) and not scope creep. As a quick _workaround_ we can simply set our own quality checker for tests, but apart from that I believe qualitycheck-related changes deserve their own discussion.

**Comment by julen** [comment]
I made a couple of changes here:
* When no snapshots are found on disk, tests will fail directly. Previously, tests were skipped and the snapshots were written to disk.
* Snapshot generation now happens upon providing the `--generate-snapshots` flag to the `py.test` command. It will collect and execute only the tests using the snapshot feature, and reference files will be created for them on disk.

**Comment by julen** [comment]
With the latest changes, now we are able to customize the output path: if a string pushed to the snapshot stack ends with a forward slash, it will be treated as a directory, and created when generating snapshots (provided it didn't exist before).

Implementation-wise, I left the responsibility to strip out or add slashes up to the caller (c.f.`as_dir()` and `url_name()` test helpers), this way the stack API surface remains minimal and simple, as there is no need to handle corner cases.

In this regards, the _example_ test function included in this PR generates snapshots at `<SNAPSHOTS_BASE_DIR>/<function-name>/<username>/<path-name>.context.json`.

With this last change I believe we are ready for merging the code in master and start using snapshot tests in other places.

File: pootle/core/utils/json.py [hunk 1]
```diff
@@ -31,17 +31,21 @@ def default(self, obj):
...
+            return super(PootleJSONEncoder, self).default(obj)
+        except TypeError:
+            return force_unicode(obj)
 
 
-def jsonify(obj):
+def jsonify(obj, indent=None):
     """Serialize Python `obj` object into a JSON string."""
-    if settings.DEBUG:
+    if settings.DEBUG and indent is not None:
```
Comment (iafan) [inline pootle/core/utils/json.py]: I believe this requires `or` rather than `and`.

**Comment by iafan** [comment]
Other than the issue reported below, gtm.

File: pootle/core/models/snapshot.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+    def __enter__(self):
+        return Snapshot(self.ctx_name, **self.snapshot_kwargs)
+
+    def __exit__(self, *args, **kwargs):
+        self.pop()
+
+    @property
+    def ctx_name(self):
+        # Disallow forward slashes
```
Comment (iafan) [inline pootle/core/models/snapshot.py]: Did you rather mean "Disallow _leading_ forward slashes"?

File: pootle/core/models/snapshot.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        stack_items = [
+            item.replace('/', '', 1) if item.startswith('/') else item
```
Comment (iafan) [inline pootle/core/models/snapshot.py]: We should be also safeguarding ourselves from paths containing relative paths like `../../`. In no circumstance the code should create directories and files outside its root snapshot directory. Probably it makes sense to fail the test if `..` is found in the file path.

File: tests/views/browse.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    '/language0/project0/subdir0/',
+    '/language0/project0/subdir0/store4.po',
+])
+def test_browse(client, request_users, test_name, flush_redis,
+                snapshot_stack, url):
+    """Tests correctness of the browsing view context."""
+    user = request_users['user']
+
+    with snapshot_stack.push([
+        as_dir(test_name), as_dir(user.username), url_name(url)
```
Comment (julen) [inline tests/views/browse.py]: No, we won't have deeply nested structures unless that's intentional (see the structure of the current test for an example). With the `url_name()` helper the slashes are converted to underscores.

File: pootle/core/models/snapshot.py (hunk 2 above)
Comment (julen) [inline pootle/core/models/snapshot.py]: Yes, good catch!

File: pootle/core/models/snapshot.py (hunk 3 above)
Comment (julen) [inline pootle/core/models/snapshot.py]: I'm leaving this responsibility out of the stack, in the end it doesn't know about any root snapshot directories. However, I'll add the check and a test in the `Snapshot` object to ensure all final paths fall below the snapshot directory.
